### ISSUE TITLE
MongoDB timeout

### ISSUE THREAD

**Issue by dvanwinkle** [ParsePlatform/parse-server#2377]
### Issue Description

I have a little over 7 million users, and when doing one of my larger queries, I am getting a MongoDB timeout. After the timeout (30 seconds), no future queries run. Everything is fine except for this one method. The query normally takes ~40-60 seconds to run. I would love to be able to refactor the collection to increase the speed of the query, but, I am running up against a deadline and need help with this timeout until I have more time. Refactoring over 7 million documents takes a decent amount of time.

I am 100% okay with the timeout, but, I am not okay with the server no longer responding to requests after the timeout. Any help would be appreciated. I am okay if there is a monitor that I can run to restart the node server or if there is something else I can do to hack together a solution before Tuesday.

### Steps to reproduce

Please include a detailed list of steps that reproduce the issue. Include curl commands when applicable.

1. Run a query that will take longer than 30 seconds
2. Receive a MongoDB timeout unhandled exception
3. Try to perform more queries

#### Expected Results

Queries after a MongoDB timeout still respond.

#### Actual Outcome

Queries after a MongoDB timeout do not respond.

### Environment Setup

- **Server**
  - parse-server version: `2.2.16`
  - Operating System:     `Ubuntu 14.04 LTS`
  - Hardware:             `Azure D15 v2` (overkill for testing move)
  - Localhost or remote server? (AWS, Heroku, Azure, Digital Ocean, etc): `Azure`

- **Database**
  - MongoDB version: `3.0.10`
  - Storage engine:  `RocksDB`
  - Hardware:        `Azure D15 v2` (different instance)
  - Localhost or remote server? (AWS, mLab, ObjectRocket, Digital Ocean, etc): `Azure`

### Logs/Trace

```
verbose: error: MongoError: server *removed*:27017 timed out
    at closeHandler (*removed*/node_modules/mongodb-core/lib/topologies/server.js:436:40)
    at emitTwo (events.js:87:13)
    at emit (events.js:172:7)
    at closeHandler (*removed*/node_modules/mongodb-core/lib/connection/pool.js:144:10)
    at g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at emit (events.js:172:7)
```

**Comment by flovilmart** [comment]
Hey, Thanks for reporting it. Starting 2.2.17 it should have a better handling of errors by recycling the connections. But I have to admit, I didn't run any tests with timeouts, mostly with lost connections. If you're able to push a PR, we'll make sure to review in priority, merge, an probably release a patch. 

Does that sound reasonable? 

As I'll be less connected next week, I'll let @tylerbrock and @drew-gross take care of you .)

**Comment by bohemima** [comment]
How does the query look like? Perhaps a better index would help?

**Comment by steven-supersolid** [comment]
Adding appropriate indexes would be the best solution but perhaps there's a reason that's not possible?

You may be able to modify the database connection settings to increase the timeout as a quick fix. I don't think this is possble with the CLI. Try modifying socketTimeoutMS by setting databaseOptions in the parameter passed to ParseServer:
```
config.databaseOptions = {
socketTimeoutMS: 60000
};
var api = new ParseServer(config);
```
Be careful doing this if you have a large number of requests because these will also time out at the new value, causing connections to be used up. You can modify your connection pool settings with the same object.

More settings here:
http://mongodb.github.io/node-mongodb-native/2.1/reference/connecting/connection-settings/

The driver should be reconnecting automatically but we can perhaps recreate the issue by using the above setting and reducing to 0.

It seems Azure can have connection issues so this could be an Azure issue. E.g. found this SO question from a while back (to do with connection timeout, not query timeout): http://stackoverflow.com/questions/25843963/mongodb-connection-timeout-when-mongodb-is-not-in-the-same-vm

**Comment by TylerBrock** [comment]
Yeah, I totally agree that the driver should reconnect automatically after the timeout. That is troubling. Let's try to get a failing test case that sets the timeout low, times out, and is unable to make further database requests.

In the meantime, I would do as @steven-supersolid and @bohemima suggest and add an index if possible.

**Comment by dvanwinkle** [comment]
The query:

```json
{
  where: {
    totalPoints: {$gt: ...}
  },
  count: ...
}
```

This is where I am trying to get the position on the leaderboard that a player is. The Query Plan is pretty straight forward and it picks a decent index if I just do a `db.getCollection('_User').explain().count({totalPoints: {$gt: 27000}})` in `mongo shell`.

```json
{
	"queryPlanner" : {
		"plannerVersion" : 1,
		"namespace" : "flinch._User",
		"indexFilterSet" : false,
		"parsedQuery" : {
			"totalPoints" : {
				"$gt" : 27000
			}
		},
		"winningPlan" : {
			"stage" : "COUNT",
			"inputStage" : {
				"stage" : "COUNT_SCAN",
				"keyPattern" : {
					"totalPoints" : -1,
					"displayname" : 1
				},
				"indexName" : "totalPoints_-1_displayname_1",
				"isMultiKey" : false
			}
		},
		"rejectedPlans" : [ ]
	},
	"serverInfo" : {
		"host" : "flinch-mongo1",
		"port" : 27017,
		"version" : "3.0.10-1.5",
		"gitVersion" : "376fb629b3911b74ee040029dd2bdc57fb2927af"
	},
	"ok" : 1
}
```

Actually running the query return in milliseconds, and returns: `2068822`.

Thanks in advance for any help!

**Comment by dvanwinkle** [comment]
I should probably note that this also runs slow on parse.com's server.

<img width="1260" alt="screenshot 2016-07-26 03 10 20" src="https://cloud.githubusercontent.com/assets/967580/17129042/b3c51242-52de-11e6-9359-186cc0b9c3b8.png">

**Comment by bohemima** [comment]
Do you run the count() as master? If not, the actual query will be something like this and most likely have a huge effect on the query `{ totalPoints: { '$gt': XXX }, _rperm: { '$in': [ null, '*' ] } }`

**Comment by dvanwinkle** [comment]
@bohemima: I didn't even think about that! I will probably just move that call to cloud code then. Thanks for the insight!

**Comment by dvanwinkle** [comment]
It seems that the restart issue has been solved in `2.2.17`. I will leave this open in case there is still a desire to fix the error issue.

**Comment by flovilmart** [comment]
Should be all fixed in 2.2.17, closing. Feel free to reopen if this closing was too hasty.
